### PR TITLE
Changes audio resource attachments to html5 audio controls

### PR DESCRIPTION
**Pull request by bryangerlach**
This will convert this:

`[Voice0053.mp3](:/e67afd4316f44704bfe10db589525910)`

to this:

Capture

which allows you to directly play the audio from within the note.

### CODE REVIEW

File: ReactNativeClient/lib/joplin-renderer/MdToHtml/rules/audio.js [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+	markdownIt.renderer.rules.link_open = (tokens, idx, options, env, self) => {
+		const Resource = ruleOptions.ResourceModel;
+
+		const token = tokens[idx];
+		const src = utils.getAttr(token.attrs, 'href');
+
+		if (!Resource.isResourceUrl(src) || ruleOptions.plainResourceRendering) return defaultRender(tokens, idx, options, env, self);
+
+		const r = utils.imageReplacement(ruleOptions.ResourceModel, src, ruleOptions.resources, ruleOptions.resourceBaseUrl);
```
Comment (CalebJohn) [inline ReactNativeClient/lib/joplin-renderer/MdToHtml/rules/audio.js:15]: Feels wrong to call imageReplacement here, this function should probably be generalized to resourceReplacement.

File: ReactNativeClient/lib/joplin-renderer/utils.js [hunk 2]
```diff
@@ -138,7 +138,7 @@ utils.imageReplacement = function(ResourceModel, src, resources, resourceBaseUrl
 	}
 
 	const mime = resource.mime ? resource.mime.toLowerCase() : '';
-	if (ResourceModel.isSupportedImageMimeType(mime)) {
+	if (ResourceModel.isSupportedImageMimeType(mime) || ResourceModel.isSupportedAudioMimeType(mime)) {
```
Comment (CalebJohn) [inline ReactNativeClient/lib/joplin-renderer/utils.js:141]: In line with the comment above, this function (`imageReplacement`) should be generalized to `resourceReplacement` and this line should be changed. In the future there may be more mimeTypes added so it would be nice to have a more flexible check here. 

Why did you have multiple function calls here instead of having one long array of supported mimeTypes?

File: ReactNativeClient/lib/joplin-renderer/MdToHtml/rules/audio.js (hunk 1 above)
Comment (CalebJohn) [reply ReactNativeClient/lib/joplin-renderer/MdToHtml/rules/audio.js:15]: Scratch that, I had a misunderstanding about what this does,

File: ReactNativeClient/lib/joplin-renderer/utils.js (hunk 2 above)
Comment (CalebJohn) [reply ReactNativeClient/lib/joplin-renderer/utils.js:141]: For example isSupportedImageMimeType could return a resource type (image, audio, video, etc.) which would then be included in the return of this function. Back in the rule you would then perform the check `if (r && r.type === "audio")` this way there would be solid support for other mimeTypes. As it stands now this code is limited to just audio (and image)

**Comment by CalebJohn** [comment]
This is really cool! It's fairly simple so there isn't too much to say but I did leave some comments. However I do think we do need to get some additional input before moving forward.

File: ReactNativeClient/lib/joplin-renderer/MdToHtml/rules/audio.js (hunk 1 above)
Comment (bryangerlach) [reply ReactNativeClient/lib/joplin-renderer/MdToHtml/rules/audio.js:15]: Actually, I think you were right at first. I had actually changed it to resourceReplacement initially and I can't remember why I changed it back.

File: ReactNativeClient/lib/joplin-renderer/utils.js (hunk 2 above)
Comment (bryangerlach) [reply ReactNativeClient/lib/joplin-renderer/utils.js:141]: The reason I have two separate function calls is because currently when you attach a file, the code uses isSupportedImageMimeType to add an exclamation mark before the resource if it is an image. So if I make the changes you are suggesting here I think that would be good. Also the logic for adding the exclamation mark would be changed to use the resource type that is returned from isSupportedMimeType.

File: ReactNativeClient/lib/joplin-renderer/utils.js [hunk 3]
```diff
@@ -138,7 +138,7 @@ utils.imageReplacement = function(ResourceModel, src, resources, resourceBaseUrl
 	}
 
 	const mime = resource.mime ? resource.mime.toLowerCase() : '';
-	if (ResourceModel.isSupportedImageMimeType(mime)) {
+	if (ResourceModel.isSupportedMimeType(mime) != 'unknown') {
 		let newSrc = `./${ResourceModel.filename(resource)}`;
 		if (resourceBaseUrl) newSrc = resourceBaseUrl + newSrc;
 		return {
```
Comment (CalebJohn) [inline ReactNativeClient/lib/joplin-renderer/utils.js:144]: If you add the result to from isSupportedMimeType to this return statement, then you'll be able to use it in the rules. This won't make your code any more functional, but it will lay the foundation for future changes of this type.

File: ReactNativeClient/lib/joplin-renderer/utils.js [hunk 4]
```diff
@@ -122,7 +122,7 @@ utils.resourceStatus = function(ResourceModel, resourceInfo) {
 	return resourceStatus;
 };
 
-utils.imageReplacement = function(ResourceModel, src, resources, resourceBaseUrl) {
+utils.resourceReplacement = function(ResourceModel, src, resources, resourceBaseUrl) {
```
Comment (CalebJohn) [inline ReactNativeClient/lib/joplin-renderer/utils.js:125]: This name change was a mistake on my part, I thought I deleted the comment but I should have left a note. Please revert this change. Sorry about that.

Comment (bryangerlach) [reply ReactNativeClient/lib/joplin-renderer/utils.js:125]: Doesn't it makes sense to call it resourceReplacement since it is being used to replace both image and audio resources with a src that can be used with HTML tags?

Comment (CalebJohn) [reply ReactNativeClient/lib/joplin-renderer/utils.js:125]: That was my initial thought as well. But now I think the name comes from the fact that this function will replace a resource with the download image. Maybe we can come up with something more clear? `notReadyResourceReplacement`? But less wordy.

Comment (bryangerlach) [reply ReactNativeClient/lib/joplin-renderer/utils.js:125]: Isn't it taking a src like `:/e67afd4316f44704bfe10db589525910` and returning something usable in html like `file://C:/Users/bryan/.config/joplindev-desktop/resources/e67afd4316f44704bfe10db589525910.mp3` ? It will only replace the resource with the download image if not ready.

Comment (CalebJohn) [reply ReactNativeClient/lib/joplin-renderer/utils.js:125]: Fair point, resourceReplacement is probably fine in that case. Let's just wait for official word on that then.

File: ReactNativeClient/lib/joplin-renderer/utils.js (hunk 3 above)
Comment (bryangerlach) [reply ReactNativeClient/lib/joplin-renderer/utils.js:145]: With isSupportedMimeType returning the type, we can basically add one line to the audio.js file to also support video files. Should this be implemented here (change the name to audiovideo.js) or have both an audio.js and video.js?

**Comment by laurent22** [comment]
Thanks @bryangerlach for the pull request and apologies for the delay reviewing it (The GSoC slowed us down quite a bit!).

You and @CalebJohn already did a good job at getting the PR ready so there's very little to change. I've only noticed these issues:

- `isSupportedMimeType` function which implies that it returns a boolean when it returns a string. Since the goal is to know the supported media type, could you rename it to `mimeTypeToMediaType()` please?
- It's not clear how it works on mobile. Did you try on it? It needs to either work the same or be disabled
- There should be a setting to turn the plugin on and off (on by default with `wysiwygNo` option)

**Comment by bryangerlach** [comment]
> * It's not clear how it works on mobile. Did you try on it? It needs to either work the same or be disabled

For me it works exactly the same on mobile as on desktop. I have only tested on Android.

Screenshot_1591648801

**Comment by laurent22** [comment]
> For me it works exactly the same on mobile as on desktop. I have only tested on Android.

Thanks for checking! Then if you could look at the few points I mentioned above and we can merge.

**Comment by bryangerlach** [comment]
I made the requested changes as well as converting to typescript.

> * `isSupportedMimeType` function which implies that it returns a boolean when it returns a string. Since the goal is to know the supported media type, could you rename it to `mimeTypeToMediaType()` please?
> * It's not clear how it works on mobile. Did you try on it? It needs to either work the same or be disabled
> * There should be a setting to turn the plugin on and off (on by default with `wysiwygNo` option)

**Comment by laurent22** [comment]
Thanks for the pull request. There are still a few changes to make, like I see you've added support for videos, so labels and identifiers should be renamed from "audio" to "media", but I will make these changes.

**Comment by laurent22** [comment]
Unfortunately I had to revert this pull request because it's now rendering note links as images: https://discourse.joplinapp.org/t/joplin-for-windows-1-0-234-bug/10558

I've reverted it in this commit: 619fa1d60780d878efee0138d4faa0db0b0e0d0b but feel free to resubmit the pull request once the link issue is fixed.
